server/v2/stf/branch/state.go

```diff
@@ -0,0 +1,81 @@
+package branch
+
+import (
+	"fmt"
+	"unsafe"
+
+	"cosmossdk.io/server/v2/core/store"
+)
+
+func NewWritersMap(
+	state store.ReaderMap,
+	branch func(readonlyState store.Reader) store.Writer) store.WriterMap {
+
+	return WritersMap{
+		state:               state,
+		branchedWriterState: make(map[string]store.Writer),
+		branch:              branch,
+	}
+}
+
+type WritersMap struct {
+	state               store.ReaderMap
+	branchedWriterState map[string]store.Writer
+	branch              func(state store.Reader) store.Writer
+}
+
+func (b WritersMap) GetReader(actor []byte) (store.Reader, error) {
+	return b.GetWriter(actor)
+}
+
+func (b WritersMap) GetWriter(actor []byte) (store.Writer, error) {
+	actorState, ok := b.branchedWriterState[unsafeString(actor)]
+	if ok {
+		return actorState, nil
+	}
+
+	writerState, err := b.state.GetReader(actor)
+	if err != nil {
+		return nil, err
+	}
+
+	actorState = b.branch(writerState)
+	b.branchedWriterState[string(actor)] = actorState
```

Comment:
do we not want to store `unsafeString(actor)` into a variable and use that instead?

Reply:
we have no way to guarantee that the lifetime of `actor []byte`  is the same as `WritersMap`.

Reply:
what unsafe string does is that it casts a []byte to string, instead of making a copy. 

If we were to use the same technique in line 43 we would end up storing an unsafe pointer to bytes as a map key. 
if that []byte gets gc'ed then the map field points to invalid memory. 

In line 32 we're only getting the value, and we're guaranteed that actor []byte will not get gc'ed before we exit the function.